[windows] integrated driver and etw based http connections.

derekwbrown opened the pull request:
Integrates driver and ETW based connections.
Always enables both driver and ETW based http monitoring when http monitoring enabled.
Honors http enabled and https enabled 
- when http enabled but not https, then ETW (which is only mode that traps https) will drop https traffic
Honors the request buffer size limit; is runtime configurable (does not use the preprocessor directive)

### Motivation

Further enables http(s) traffic monitoring for NPM

### Additional Notes


### Possible Drawbacks / Trade-offs

### Describe how to test/QA your changes

(All on a windows system)
Enable NPM driver
Install IIS
Install some other web server (doesn't matter what).

Ensure that, when http tracking enabled, that http traffic is recorded by system probe.
Traffic handled by IIS should have the `dynamicTags` present; all other traffic should not.
Ensure that https traffic is _not_ recorded when https is not enabled.

Re run with https enabled.
Ensure that https traffic to IIS is tracked accordingly (with dynamic tags), other traffic is handled as before.

In both cases, ensure that http client traffic is also recorded.  Note that in all cases https client traffic will not
be recorded (expected behavior)

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

derekwbrown committed:
Update to latest driver version

derekwbrown committed:
Checkpoint
Seems to be correctly disambiguating etw & driver based flows

Add ability to exclude https in etw

Suggestions for db (#13480)

Finalize ETW/driver integration
Make sure that request buffer config is honored.
Adds dynamic tag for site id

derekwbrown committed:
gofmt

derekwbrown committed:
fix my linux mistakes

derekwbrown committed:
Remove unused code

iglendd commented on pkg/network/http/http_statkeeper_windows.go:
```diff
@@ -0,0 +1,17 @@
+)
+
+func getPathBufferSize(c *config.Config) int {
+	return int(c.HTTPMaxRequestFragment)
```
I see why you have added this. But I wonder if existing global config alone is sufficient

derekwbrown replied:
again it's runtime configurable.

derekwbrown committed:
add service tag

derekwbrown commented on pkg/network/etw/etw-http-service.go:
```diff
@@ -160,7 +160,7 @@ type Http struct {
 	//
 	// Url           string
-	// SiteID        uint32
+	SiteID uint32
```
we wer having conversations about whether to use that as part of the service tag.
I also found out how to use the config and map the site ID to name (for a future PR).  Right now it's just in there "to have"

derekwbrown commented on pkg/network/etw/etw-http-service.go:
```diff
@@ -641,6 +675,7 @@ func httpCallbackOnHTTPRequestTraceTaskParse(eventInfo *C.DD_ETW_EVENT_INFO) {
 		// in future it probably should be optimize because we have already
 		// whole thing
 		httpConnLink.http.RequestFragment = make([]byte, maxRequestFragmentBytes)
+		httpConnLink.http.Txn.MaxRequestFragment = uint16(maxRequestFragmentBytes)
```
So there's a couple of things here that need to be unpacked.
The intention of this was to have it be the actual length so that only the actual length was copied if the fragment was less than the max.
Also, the driver is runtime configurable so it could, in theory, change.

derekwbrown commented on pkg/network/etw/etw-utils.go:
```diff
@@ -261,7 +261,7 @@ func parseUnicodeString(data []byte, offset int) (val string, nextOffset int, va
 		lenString = len(data) - offset
 	} else {
-		lenString = termZeroIdx - offset
+		lenString = termZeroIdx
```
I changed b/c I was debugging and it was cutting off the string at the wrong place.
If you look at it `term0idx` is already zero based from offste (because the call to `bytesOfDoubleZero` starts at `offset:`.  So the old code ended up effectively subtracting `offset` twice.

iglendd replied:
damn you are right, I did not notice it

derekwbrown committed:
Merge branch 'main' into db/normalize-driver-etw

derekwbrown committed:
remove BOM from types-windows, and fix generating script

derekwbrown committed:
re-add types-windows without BOM

derekwbrown committed:
Fix statkeeper tests on Windows
needs a fully defined config because request fragment size is runtime configurable
